File: luogu/p1219.py
n=int(input())
lie=[]
ans=[] 
for i in range(n+1):
    lie.append(0)
    ans.append(0)
zx=[]
yx=[]
for i in range(61):
    zx.append(0)
    yx.append(0)
c=0
def dfs(t):
    global n,c,ans,zx,yx,lie
    if t==n+1:
        c+=1
        if c<=3:
            for i in range(1,n+1):
                print(ans[i],end=' ')
            print("")
    else:
        for i in range(1,n+1):
            if lie[i]==0 and zx[t+i]==0 and yx[t-i+30]==0:
                lie[i]=1 
                zx[t+i]=1
                yx[t-i+30]=1
                ans[t]=i
                dfs(t+1)
                lie[i]=0 
                zx[t+i]=0
                yx[t-i+30]=0
                ans[t]=0
dfs(1)
print(c)

File: luogu/cpp/p1219.cpp
#include<cstdio>
int x[14],y[60],z[60],n,ans,anss[14];
void dfs(int l){
	if(l==n+1){
		ans++;
		if(ans<=3){
			for(int i=1;i<=n;i++){
				printf("%d ",anss[i]);
			}
			printf("\n");
		}
	}
	else
	{
		for(int i=1;i<=n;i++)
		{
			if(x[i]==0&&y[i-l+30]==0&&z[i+l]==0){
				anss[l]=i;
				x[i]=1;
				y[i-l+30]=1;
				z[i+l]=1;
				dfs(l+1);
				x[i]=0;
				y[i-l+30]=0;
				z[i+l]=0;
			}
		}
	}
}
int main(){
	scanf("%d",&n);
	dfs(1);
	printf("%d",ans);
	return 0;
} 
